howto_io.md

```diff
@@ -0,0 +1,268 @@
+$ cat -T tabs.txt
+1^I2^I3
+44^I ^I6
+7^I888^I9
+
+>>> np.genfromtxt("tabs.txt",delimiter="\t",missing_values=" +")
+array([[  1.,   2.,   3.],
+       [ 44.,  nan,   6.],
+       [  7., 888.,   9.]])
+```
+
+## Read a file in .npy or .npz format
+
+Use [numpy.load](https://numpy.org/doc/stable/reference/generated/numpy.load.html).
+
+## Write to a file to be read back by NumPy
+### Binary
+Use [numpy.save](https://numpy.org/doc/stable/reference/generated/numpy.save.html) to store a single array,
+[numpy.savez](https://numpy.org/doc/stable/reference/generated/numpy.savez.html) or
+[numpy.savez_compressed](https://numpy.org/doc/stable/reference/generated/numpy.savez_compressed.html#numpy.savez_compressed)
+to store multiple arrays. For [security and portability](#saverestore-using-a-pickle-file), set `allow_pickle=False`.
+
+Masked arrays [can't currently be saved](https://numpy.org/devdocs/reference/generated/numpy.ma.MaskedArray.tofile.html),
+nor can other arbitrary array subclasses.
+
+### Human-readable
+`save` and `savez` create binary files. To write a human-readable file, use
+[numpy.savetxt](https://numpy.org/doc/stable/reference/generated/numpy.savetxt.html).
+The array can only be 1- or 2-dimensional, and there's no `savetxtz` for multiple files.
+
+### Large arrays
+See [Write or read large arrays](#large_arrays).
+
+## Read an arbitrarily formatted binary file ("binary blob")
+
+Use a [structured array](https://numpy.org/doc/stable/user/basics.rec.html).
+
+**Example:**
+
+The `.wav` file header is a 44-byte block preceding `data_size` bytes of the
+actual sound data:
+
+```
+chunk_id         "RIFF"
+chunk_size       4-byte unsigned little-endian integer
+format           "WAVE"
+fmt_id           "fmt "
+fmt_size         4-byte unsigned little-endian integer
+audio_fmt        2-byte unsigned little-endian integer
+num_channels     2-byte unsigned little-endian integer
+sample_rate      4-byte unsigned little-endian integer
+byte_rate        4-byte unsigned little-endian integer
+block_align      2-byte unsigned little-endian integer
+bits_per_sample  2-byte unsigned little-endian integer
+data_id          "data"
+data_size        4-byte unsigned little-endian integer
+```
+The `.wav` file header as a NumPy structured dtype:
+```
+wav_header_dtype = np.dtype([
+    ("chunk_id", (bytes, 4)), # flexible-sized scalar type, item size 4
+    ("chunk_size", "<u4"),    # little-endian unsigned 32-bit integer
+    ("format", "S4"),         # 4-byte string
+    ("fmt_id", "S4"),
+    ("fmt_size", "<u4"),
+    ("audio_fmt", "<u2"),     #
+    ("num_channels", "<u2"),  # .. more of the same ...
+    ("sample_rate", "<u4"),   #
+    ("byte_rate", "<u4"),
+    ("block_align", "<u2"),
+    ("bits_per_sample", "<u2"),
+    ("data_id", "S4"),
+    ("data_size", "u4"),
+    #
+    # the sound data itself cannot be represented here:
+    # it does not have a fixed size
+   ])
+
+data = np.load(f,dtype=wave_header_dtype)

```

Comment:
The example (at the link http://scipy-lectures.org/advanced/advanced_numpy/index.html#example-reading-wav-files) actually uses np.fromfile, no?

Reply:
Thanks -- the tutorial focused on the dtype and didn't actually show the read routine, and I supplied the wrong one. Fixed.